slashing/program/src/shred.rs

```diff
@@ -0,0 +1,567 @@
+const SIZE_OF_PARENT_OFFSET: usize = 2;
+const SIZE_OF_NUM_DATA_SHREDS: usize = 2;
+const SIZE_OF_NUM_CODING_SHREDS: usize = 2;
+const SIZE_OF_POSITION: usize = 2;
+
+const SIZE_OF_MERKLE_ROOT: usize = 32;
+const SIZE_OF_MERKLE_PROOF_ENTRY: usize = 20;
+
+const OFFSET_OF_SHRED_VARIANT: usize = SIZE_OF_SIGNATURE;
+const OFFSET_OF_SLOT: usize = SIZE_OF_SIGNATURE + SIZE_OF_SHRED_VARIANT;
+const OFFSET_OF_INDEX: usize = OFFSET_OF_SLOT + SIZE_OF_SLOT;
+const OFFSET_OF_VERSION: usize = OFFSET_OF_INDEX + SIZE_OF_INDEX;
+const OFFSET_OF_FEC_SET_INDEX: usize = OFFSET_OF_VERSION + SIZE_OF_VERSION;
+
+const OFFSET_OF_DATA_PARENT_OFFSET: usize = OFFSET_OF_FEC_SET_INDEX + SIZE_OF_FEC_SET_INDEX;
+const OFFSET_OF_DATA_SHRED_FLAGS: usize = OFFSET_OF_DATA_PARENT_OFFSET + SIZE_OF_PARENT_OFFSET;
+
+const OFFSET_OF_CODING_NUM_DATA_SHREDS: usize = OFFSET_OF_FEC_SET_INDEX + SIZE_OF_FEC_SET_INDEX;
+const OFFSET_OF_CODING_NUM_CODING_SHREDS: usize =
+    OFFSET_OF_CODING_NUM_DATA_SHREDS + SIZE_OF_NUM_DATA_SHREDS;
+const OFFSET_OF_CODING_POSITION: usize =
+    OFFSET_OF_CODING_NUM_CODING_SHREDS + SIZE_OF_NUM_CODING_SHREDS;
+
+type MerkleProofEntry = [u8; 20];
+const MERKLE_HASH_PREFIX_LEAF: &[u8] = b"\x00SOLANA_MERKLE_SHREDS_LEAF";
+const MERKLE_HASH_PREFIX_NODE: &[u8] = b"\x01SOLANA_MERKLE_SHREDS_NODE";
+
+#[repr(transparent)]
+#[derive(Clone, Copy, Default, Eq, PartialEq, Ord, PartialOrd, Hash, Deserialize)]
+pub(crate) struct Signature(GenericArray<u8, U64>);
+
+bitflags! {
+    #[derive(Clone, Copy, Debug, Default, Eq, PartialEq, Deserialize)]
+    pub struct ShredFlags:u8 {
+        const SHRED_TICK_REFERENCE_MASK = 0b0011_1111;
+        const DATA_COMPLETE_SHRED       = 0b0100_0000;
+        const LAST_SHRED_IN_SLOT        = 0b1100_0000;
+    }
+}
+
+#[repr(u8)]
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq, IntoPrimitive, TryFromPrimitive)]
+pub(crate) enum ShredType {
+    Data = 0b1010_0101,
+    Code = 0b0101_1010,
+}
+
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+enum ShredVariant {
+    LegacyCode,
+    LegacyData,
+    MerkleCode {
+        proof_size: u8,
+        chained: bool,
+        resigned: bool,
+    },
+    MerkleData {
+        proof_size: u8,
+        chained: bool,
+        resigned: bool,
+    },
+}
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub(crate) struct ErasureMeta {
+    num_data_shreds: usize,
+    num_coding_shreds: usize,
+    first_coding_index: u32,
+}
+
+#[derive(Clone, PartialEq, Eq)]
+pub(crate) struct Shred {
+    shred_type: ShredType,
+    proof_size: u8,
+    chained: bool,
+    resigned: bool,
+    payload: Vec<u8>,
+}
+
+impl Shred {
+    const SIZE_OF_CODING_PAYLOAD: usize = 1228;
+    const SIZE_OF_DATA_PAYLOAD: usize =
+        Self::SIZE_OF_CODING_PAYLOAD - Self::SIZE_OF_CODING_HEADERS + SIZE_OF_SIGNATURE;
+    const SIZE_OF_CODING_HEADERS: usize = 89;
+    const SIZE_OF_DATA_HEADERS: usize = 88;
+
+    pub(crate) fn new_from_payload(payload: Vec<u8>) -> Result<Self, SlashingError> {
```

Comment:
Not quite a Pod since I stored some of the commonly used fields `shred_type, proof_size` and the shred format flags in the struct. Wanted to avoid deserializing them multiple times.

But in the spirit of your suggestion I changed it to accept a pointer so that we can avoid copying the payload. We'll ultimately need to copy in order to compare the fields when checking the proof, but this avoids copying unnecessary bits.